src/Orleans/Streams/Internal/StreamConsumerExtension.cs

```diff
@@ -63,50 +67,90 @@ internal StreamSubscriptionHandleImpl<T> SetObserver<T>(GuidId subscriptionId, S
             }
         }
 
-        internal bool RemoveObserver(GuidId subscriptionId)
+        public Task OnSubscriptionChange<T>(Func<StreamSubscriptionHandle<T>, Task> onAdd, Func<string, IStreamIdentity, Guid, Task> onRemove = null)
+        {
+            this.onSubscriptionAddAction = onAdd;
```

Comment:
This is another case where having a typed handle (there have been others) creates problems.  Raising the question, do we want to bit the bullet and change the handle to not be typed.
If we keep the handle typed (probably best for now) I see two problems with this:
1) What happens if someone needs to process streams of different types, they only get one callback?
2) By storing these actions this way, it creates the need to do the reflection magic later, on each call.
I suggest creating an abstraction, something like ISubscriptionChangeHandler, which have onAdd and OnRemove signatures.  Create a generic implementation of this for type specific actions, then add a map of the type to the ISubscriptionChangeHandler for each type specific actions registered.  When we have a subscription change we call the OnAdd or OnRemove on all of the registered handlers.